### PR TITLE
Fix IPC piping readable for mac CLI for larger compiled graphs

### PR DESCRIPTION
**Pull request by Ekrekr**
Fixes https://github.com/dataform-co/dataform-co/issues/9099

### CODE REVIEW

File: api/commands/compile.ts [hunk 1]
```diff
@@ -97,10 +97,14 @@ export class CompileChildProcess {
       const pipe = this.childProcess.stdio[4] as Readable;
       const chunks: Buffer[] = [];
       pipe?.on("readable", () => {
-        let buffer: Buffer = pipe.read();
-        while (buffer) {
-          chunks.push(buffer);
+        let firstRead = true;
+        let buffer: Buffer;
+        while (buffer || firstRead) {
```
Comment (Ekrekr) [inline api/commands/compile.ts:102]: Assignment in while loop not allowed, so can't do `while(buffer = pipe.read())`

File: api/vm/compile.ts [hunk 2]
```diff
@@ -69,9 +69,20 @@ export function compile(compileConfig: dataform.ICompileConfig) {
...
+      let splitCompiledResult: string[] = [];
+      while (compiledResult.length > 0) {
+        splitCompiledResult.push(compiledResult.substring(0, 2 ** 12));
+        compiledResult = compiledResult.substring(2 ** 12);
+      }
+      for (let chunk of splitCompiledResult) {
+        writer.write(chunk, "utf8");
+        // This is needed for an unclear reason.
+        // tslint:disable-next-line: no-console
+        console.log();
```
Comment (Ekrekr) [inline api/vm/compile.ts:84]: Any ideas why?

File: api/commands/compile.ts [hunk 3]
```diff
@@ -97,10 +97,14 @@ export class CompileChildProcess {
...
+          firstRead = false;
           buffer = pipe.read();
+          if (!!buffer) {
+            chunks.push(buffer);
+          }
```
Comment (BenBirt) [inline api/commands/compile.ts:107]:
```suggestion
        let buffer: Buffer = pipe.read();
        while (buffer) {
          chunks.push(buffer);
          buffer = pipe.read();
        }
```

File: api/vm/compile.ts (hunk 2 above)
Comment (BenBirt) [reply api/vm/compile.ts:84]: are you sure the actual problem isn't that you're just not waiting for the write to complete? `socket.write` takes a callback which is called when the write completes, seems like you should `await` / `promisify`: https://nodejs.org/api/net.html#net_socket_write_data_encoding_callback

File: api/commands/compile.ts (hunk 3 above)
Comment (Ekrekr) [reply api/commands/compile.ts:107]: Reading when there's nothing in the pipe throws an error; the second read can read when there's only one thing to read

**Comment by Ekrekr** [comment]
Yup I was entirely wrong about the limit, can just pass the process exit in the callback 🤷‍♂️

File: api/vm/compile.ts [hunk 4]
```diff
@@ -71,15 +71,14 @@ export function listenForCompileRequest() {
     try {
       const compiledResult = compile(compileConfig);
       const writer = new net.Socket({ fd: 4 });
-      writer.write(compiledResult);
+      writer.write(compiledResult, () => process.exit());
```
Comment (BenBirt) [inline api/vm/compile.ts:74]: i think we should avoid this, it feels like a code smell. just `await` `promisify`?

File: api/commands/compile.ts (hunk 3 above)
Comment (BenBirt) [reply api/commands/compile.ts:107]: i think my code is equivalent to yours

File: api/vm/compile.ts (hunk 4 above)
Comment (Ekrekr) [reply api/vm/compile.ts]: Promisify doesn't seem to be valid on `writer.write`... or in any case I couldn't get it to work. How about this alternative?

**Review by BenBirt — APPROVED** [review]

**Pull request merged**